docs/BUILDING.md

```diff
@@ -122,6 +123,92 @@ If you want to run the quick-lint-js program, you can find it in
 `build\Debug\quick-lint-js.exe`. Run it from a command prompt or PowerShell
 window.
 
+---
+
+### macOS and Linux: nix
+
+This is for advanced users only. You need to
+[install the Nix package manager](https://nixos.org/download.html) if you have
+not already done so.
+
+[Nix][] uses derivation files that contain instructions for automation on how
+to build and package software. The `dist/` folder contains such derivations for
+quick-lint-js and can be used to download dependencies and build it.
+
+#### 1. Downloading build dependencies
+
+The derivations in `dist/` contain the necessary dependencies to build
+quick-lint-js using cmake and ninja. We can use `nix-shell` to download those
+dependencies and make them available for development.
+
+    $ nix-shell dist/shell.nix
+
+While it seems this does not do anything, you will be dropped in a shell that
+has the build tools available necessary to build quick-lint-js. For example,
+you can use `cmake` now, even if it is not installed on your system.
+
+Note that once you exit this shell, the downloaded dependencies will not be
+available anymore. After a while, they might be deleted from your system by
+`nix`'s garbage collection to free up disk space. You can use the above command
+again to re-create a shell with the build tools available at any time.
+
+By default, the above will set the `CMAKE_BUILD_TYPE` to be Debug. If you
+want to build the release version, use the following instead:
+
+    $ nix-shell dist/shell.nix --argstr buildType Release
+
+As a side note, this default only applies to `nix-shell`. If you use
+`nix-build` as described in `INSTALLING`, the default is Release. You can
+stop the tests from being built by adding `--arg test false` to the above
+command line.
+
+#### 2. Configuring
+
+After running above command, additionally to build tools like `cmake`, nix
+also prepares some shell functions for you to run the configuration, build
+and test steps.
+
+To configure the build environment, we use
+
+    $ cmakeConfigurePhase
+
+This will create a `build/` folder in the project folder and changes the
+current directory to it.
```

Comment:
I have not found a good way, unfortunately - unless we want to mangle with nix's build phases (we'd also have to introduce a cd into build in the preBuild hook, for example). We could introduce `pushd build` and `popd` to all the pre* phases, but then the commands only work when in the project folder and would not work when in the build folder (like they normally would). Or we do some more fancy "find the build folder regardless of where you are"-magic but I really don't want to do that.

The alternative to using the nix phases would be to just use nix-shell to provide building dependencies automatically and tell the user to run the same commands as in the other setups (e.g. `ninja -B build` directly instead of `ninjaBuildPhase`).